Transactions should survive graceful leadership transfers of data partitions

bharathv opened the pull request:
## Cover letter

Currently if there is a leadership transfer of data partitions, all the in-flight (non committed/preprared/aborted) transactions fail when the next set of requests are redirected to the new leader. This happens due to some volatile transaction state in rm_stm that is not actively check-pointed to the log. This PR does exactly that.

When there is a request to transfer leadership, we serialize the in memory state of transactions in flight and replicate that to the followers which then apply() that to their local state via stm upcall.

Notes

* Introduces two new transaction batch types `txn_checkpoint` and `txn_checkpoint_applied`. `txn_checkpoint` batch includes the serialized check-pointed state as payload whereas `txn_checkpoint_applied` notifies the new followers to that the leader has applied the state locally and they can choose to purge their local state if they see this batch. We need the latter because the new leader's check-pointed state can potentially diverge with the incoming transactions (we only need one source of truth, which is the leader).
* `txn_checkpoint batch` is written with `quorum_ack` and this is critical for correctness. This ensures that all the inflight requests (acks=1, preparing/commits etc) are replicated to all the nodes.
* There is a similar orthogonal problem with transaction coordinator (`tm_stm`) leadership changes that will be fixed separately.
* We can still run into issues like node crashes where this state is lost immediately. However that is not a correctness issue as the the clients retry the failed transactions when they hit the new leader with a clean state.
* Idempotent transactions can survive leadership transfers if the batch with quorum ack succeeds (state reconciled from _log_state on the new leader) and we don't need to do anything special.


Example. Say there are 3 nodes 0 (leader), 1, 2 with one data partition each.

1. 0 gets the request to transfer leadership.
2. 0 writes a txn_checkpoint batch with quorum_ack
3. 1 & 2 apply the check-pointed state to their local state machines
4. 0 goes thru with the leadership transfer and the quorum votes for node 1 as the new leader.
5. 1 upon becoming the leader issues a txn_checkpoint_applied request _before_ accepting any requests.
6. This makes sure that the new followers (0, 2) purge their local mem_state and there is only one source of truth which the current leader.
7. If there is new request to leadership transfer, just back to step (1) with node 1 writing the checkpoint batch.


Note: 

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #6213

## Release notes

* TBD

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -792,11 +792,9 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate(
-            [u = std::move(unit)] {});
-      });
+    auto units = co_await _state_lock.hold_write_lock();
+    co_await checkpoint_in_memory_state();
```
if the checkout or do_transfer_leadership fails we need explicitly step down to avoid writing requests after the checkpoint

rystsov commented on tools/offline_log_viewer/storage.py:
```diff
@@ -133,6 +133,7 @@ class BatchType(Enum):
     archival_metadata = 19
     cluster_config_cmd = 20
     feature_update = 21
+    tx_checkpoint = 22
```
We should have two new batches e.g. `send_tx_leadership` & `receive_tx_leadership`. When we reuse the same batch we overload its meaning depending on the context. It's harder to describe what it does and it may lead to situation when we invoke one code path instead of another. With complex code it's safer to be verbose rather than smart.

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -469,6 +475,11 @@ ss::future<tx_errc> rm_stm::do_prepare_tx(
     co_return tx_errc::none;
 }
 
+void rm_stm::on_leadership_change(
```
We have two source of where the modifications of rm_stm may come from:
  - rpc requests from kafka clients or txn coordinator
  - from the apply caused by replaying the log

`on_leadership_change` adds a third source of the changes and it forces us to think about more of combinations. For example, can a request from a client of a txn coordinator come before on_leadership_change is triggered, what happens if on_leadership_change fails.

Let's use sync to avoid adding a new entity. Sync guarantees that all records written by previous leader are already applied. So an apply can process `send_tx_leadership` & `receive_tx_leadership` as it goes. When sync sees unbalanced `send_tx_leadership` it should write `receive_tx_leadership` before yielding control flow. Also several syncs may be running in parallel so once it sees unbalanced `send_tx_leadership` it should use mutex to guarantee that only one `receive_tx_leadership` is written

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1827,18 +1851,34 @@ ss::future<model::record_batch> rm_stm::make_checkpoint_batch(
+ss::future<model::record_batch>
+rm_stm::make_checkpoint_purge_batch(model::term_id term) const {
+    iobuf key;
+    co_await serde::write_async(key, term);
```
term is enough, we don't need to write node_id

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -416,20 +416,43 @@ ss::future<tx_errc> rm_stm::do_prepare_tx(
+        // -----
+
+
+        if (synced_term != etag) {
```
I'm pretty sure you're right but it's dangerous to relax conditions before the release. We should  update mem_state.expected to store a term (actual at the begin_tx moment) along the tx_seq. Then when we transfer the state we'll be able to continue using the check for validating the prepare and then using the term of `receive_tx_leadership` for conditional replicate.

bharathv replied:
Redid this by tracking the term in begin_tx. Although I'm not fully convinced that version is any better than commenting the check. Anyway check the code and we can discuss.

rystsov replied:
I on the same page with you :)

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1495,7 +1495,7 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
     if (ready) {
         if (_mem_state.term != _insync_term) {
-            _mem_state = mem_state{.term = _insync_term};
+            _mem_state.term = _insync_term;
```
no-no-no

imagine leadership bounced back and forth and we updated the term but haven't reset all the other fields, we must reset them

when we replay the log we should cache the `send_tx_leadership` info and during the sync if there is unbalanced `send_tx_leadership` we should first write `receive_tx_leadership` and then update _mem_state. the default mem_state logic should not change

bharathv replied:
This was just a hack to get it working, polished most of this, have a look at the next rev.

bharathv reviewed (commented):
Will update the next rev shortly.

bharathv commented:
Latest [force-push](https://github.com/redpanda-data/redpanda/compare/faa7395b83d61a4bfae2e4e3afc9e7607f4ba24e..c3114865eeccd568a5e12951610e756145598e6b) addresses all the comments so far (I think). Still a lot of tests need to be written (ex: units around serde, expiration tracking after failover, more edge cases around failures etc). Pushing this out while I work on them.

rystsov commented on src/v/cluster/rm_stm.h:
```diff
@@ -371,7 +371,9 @@ class rm_stm final : public persisted_stm {
     };
 
-    struct mem_state {
+    struct mem_state
```
let's use data transfer object to transfer the state to the followers instead of serializing mem_state. It creates a coupling between in memory state and on disk state and makes it very hard to evolve the format independently

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1437,6 +1437,12 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
+            // All other replicate ops need to wait until this finishes, this is
+            // so that the followers clean up the state before we start applying
+            // changes to the in memory state.
+            co_await _wait_for_checkpoint_purge.with(
```
we should write tx_checkpoint_purge only after tx_checkpoint if we write it on every election it increases disk pressure and makes it harder to debug because we lose the one tx_checkpoint - one tx_checkpoint_purge invariant.

by the way let's rename tx_checkpoint and tx_checkpoint_purge to send_tx_leadership and receive_tx_leadership to highlight meaning and the relation between them

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1808,7 +1808,58 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
+
+ss::future<> rm_stm::checkpoint_in_memory_state() {
+    absl::flat_hash_map<model::producer_identity, int32_t> pid_to_tail_seq;
+    for (auto& [pid, reqs] : _inflight_requests) {
```
we don't need to pass the _inflight_requests requests, they are part of idempotency not transactions

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1808,7 +1808,58 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
+        pid_to_tail_seq.emplace(pid, reqs->tail_seq);
+    }
+    absl::flat_hash_map<model::producer_identity, int64_t> pid_to_expiry;
+    for (auto& [pid, exp] : _mem_state.expiration) {
```
we already serialize _mem_state; why do we need to serialize expiration once again?

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1710,7 +1710,7 @@ ss::future<> rm_stm::apply(model::record_batch b) {
         }
     } else if (hdr.type == model::record_batch_type::tx_checkpoint) {
-        apply_checkpoint(b);
+        apply_checkpoint(std::move(b));
```
Let's cache it instead of applying right away. in this case the sync will have a chance to reconcile the data before putting it into action instead of applying and reconciling them after the fact. The latter is more error prone.

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1711,6 +1711,8 @@ ss::future<> rm_stm::apply(model::record_batch b) {
     } else if (hdr.type == model::record_batch_type::tx_checkpoint) {
         apply_checkpoint(std::move(b));
+    } else if (hdr.type == model::record_batch_type::tx_checkpoint_purge) {
+        apply_checkpoint_purge(std::move(b));
```
same thing there, let the purge reset the cache instead of tinkering with the hot data

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1827,6 +1829,26 @@ void rm_stm::apply_checkpoint(const model::record_batch& batch) {
+
+    auto term = serde::from_iobuf<model::term_id>(std::move(key_buf));
+
+    auto am_leader = term == _c->term() && _c->is_leader();
```
don't need this check, the current leader may just take the _parked_checkpointed_mem_state before issuing the purge command so this logic will be simpler, just reset the parked state

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1900,6 +1901,34 @@ ss::future<> rm_stm::replicate_checkpoint_purge(model::term_id term) {
+        vlog(clusterlog.debug, "Tracking tx {} ms_left {}", k, ms_left.count());
+        track_tx(k, ms_left);
+    }
+    for (auto& [pid, seq] : _parked_checkpointed_mem_state._pid_to_tail_seq) {
+        _inflight_requests.emplace(pid, ss::make_lw_shared<inflight_requests>());
+        _inflight_requests[pid]->tail_seq = seq;
+        vlog(clusterlog.debug, "Tracking pid {}, tail_seq {}", pid, seq);
+    }
```
we don't need to touch _inflight_requests

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1136,19 +1136,6 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate_seq(
         co_return errc::not_leader;
     }
 
-    if (session->term < synced_term) {
```
we don't need to touch the inflight requests aren't related to transactions

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1847,6 +1834,8 @@ void rm_stm::apply_checkpoint_purge(const model::record_batch& batch) {
         return;
     }
     _parked_checkpointed_mem_state = {};
+    _mem_state = {};
```
so far only sync reset mem_state, let keep it this way

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -324,7 +324,8 @@ ss::future<checked<model::term_id, tx_errc>> rm_stm::do_begin_tx(
 
-    auto [_, inserted] = _mem_state.expected.emplace(pid, tx_seq);
+    auto [_, inserted] = _mem_state.expected.emplace(
+      pid, std::make_pair(synced_term, tx_seq));
```
lets use a struct instead of a tuple, it's easier to use names instead of .first &.second

rystsov commented on tests/rptest/tests/transactions_test.py:
```diff
@@ -244,3 +247,83 @@ def change_static_member_test(self):
+        })
+        try:
+            consumer.subscribe([self.input_t])
+            records = self.consume(consumer, max_records=count, timeout_s=10)
```
consume may return less data than requested, it's better to keep reading until we see all data or reach timeout

bharathv commented:
Rebased on dev tip, ready for review.

VadimPlh commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1502,6 +1502,12 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
+            // so that the followers clean up the state before we start applying
+            // changes to the in memory state.
+            co_await _checkpoint_applied_replicate.with(
+              [this]() { return replicate_checkpoint_applied(_insync_term); });
```
Should we need to use new feature here? 
You introduced new record type. So you should protect upgrade process

dotnwat replied:
good point. i wonder if old nodes ignore unknown comamnds? probably not

VadimPlh commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1962,6 +1966,25 @@ ss::future<> rm_stm::replicate_checkpoint_applied(model::term_id term) {
 }
 
+void rm_stm::reconcile_mem_state() {
+    if (!_c->is_leader()) return;
```
Please, use 
```
if (smth) {
}
```

rystsov replied:
this check isn't necessary we already checked with `if (ready) {` up the stack.

dotnwat commented on src/v/cluster/rm_stm.h:
```diff
@@ -205,6 +209,14 @@ class rm_stm final : public persisted_stm {
             return is_expiration_requested || deadline() <= now;
         }
+
+        void serde_write(iobuf& out) { serde::write(out, timeout); }
+
+        void serde_read(iobuf_parser& in, const serde::header& h) {
+            last_update = time_point_type::clock::now();
+            timeout = serde::read_nested<duration_type>(
+              in, h._bytes_left_limit);
+        }
```
can you use `serde_fields`?

dotnwat commented on src/v/cluster/rm_stm.h:
```diff
@@ -421,6 +435,18 @@ class rm_stm final : public persisted_stm {
             }
         }
+
+        auto serde_fields() {
+            return std::tie(
```
add to compat framework? (maybe it's in a later commit)

dotnwat commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1886,10 +1886,36 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
+    iobuf key;
+    co_await serde::write_async(key, _insync_term);
+    iobuf val;
+    co_await serde::write_async(val, std::move(_mem_state));
```
```
ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
...
    co_await serde::write_async(val, std::move(_mem_state));

```
i think that it needs to be formalized in some way that taking a checkpoint may or may not destroy _mem_state content. that doesn't seem like a casual side effect of building a checkpoint.

dotnwat commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1886,10 +1886,36 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
+
+ss::future<> rm_stm::checkpoint_in_memory_state() {
+    auto batch = co_await make_checkpoint_batch();
+    _mem_state = {}; // reset
```
it seems like it might be easier to understand if the calling context explicitly took ownership of _mem_state content and passed it into checkpointing routines so that clearing that structure isn't a side effect of data serialization / replication?

dotnwat commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1769,7 +1769,7 @@ ss::future<> rm_stm::apply(model::record_batch b) {
         }
     } else if (hdr.type == model::record_batch_type::tx_checkpoint) {
-        apply_checkpoint(b);
+        apply_checkpoint(std::move(b));
```
what is the purpose of moving `b` if `apply_checkpoint` takes a const-ref?

dotnwat commented on src/v/cluster/rm_stm.cc:
```diff
@@ -355,7 +355,10 @@ ss::future<checked<model::term_id, tx_errc>> rm_stm::do_begin_tx(
-    auto [_, inserted] = _mem_state.expected.emplace(pid, tx_seq);
+    auto [_, inserted] = _mem_state.expected.emplace(
+      pid,
+      (struct tx_inflight_info){
```
prefer the syntax tx_inflight_info{...}

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1886,10 +1886,36 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
+    if (!result) {
+        // This is on a best effort basis, if we do not replicate, txns are
+        // aborted on the new leader and the client is expected to retry.
+        vlog(clusterlog.warn, "Error replicating checkpoint batch: {}", result.error());
```
we can't eat error there, we should return an error and let code [there](https://github.com/redpanda-data/redpanda/pull/5366/commits/81d9438101aac5ecab2a28d303b741ab277c4898#r915125406) to step down

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1886,10 +1886,36 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
-ss::future<> rm_stm::checkpoint_in_memory_state() const { co_return; }
+ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
+    iobuf key;
+    co_await serde::write_async(key, _insync_term);
```
we should use the same key for the records to allow compaction to take care about the garbage

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1914,6 +1920,41 @@ ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
+ss::future<model::record_batch>
+rm_stm::make_checkpoint_applied_batch(model::term_id term) {
+    iobuf key;
+    co_await serde::write_async(key, term);
```
we should use the same key and put `term` into the value to rely on compaction

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1914,6 +1920,41 @@ ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
+}
+
+ss::future<> rm_stm::replicate_checkpoint_applied(model::term_id term) {
+    if (!_parked_checkpointed_mem_state) {
```
let's move the check a level up or rename `replicate_checkpoint_applied` to `maybe_replicate_checkpoint_applied`

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1914,6 +1920,41 @@ ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
+      std::move(reader),
+      raft::replicate_options(raft::consistency_level::quorum_ack));
+    if (!result) {
+        vassert(
```
vassert is too much, it will crash the node, it's better to step down on error

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1502,13 +1502,17 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
+            // applied before we
+            //    allow clients to progress.
+            co_await _checkpoint_applied_replicate.with([this]() {
+                reconcile_mem_state();
```
yeah, it's better to put the check here to avoid to checking it both in `reconcile_mem_state` and `replicate_checkpoint_applied`:

```
if (_parked_checkpointed_mem_state) {
    reconcile_mem_state();
    co_await replicate_checkpoint_applied()
}
```

It reduces the combination of branches from 4 to 2 so it's easier to understand what's happening.

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1990,6 +1990,15 @@ void rm_stm::reconcile_mem_state() {
 }
 
 ss::future<> rm_stm::checkpoint_in_memory_state() {
+    if (unlikely(_mem_state.all_txns_count() > checkpoint_txns_limit)) {
```
We either need to rename the method to `maybe_checkpoint_in_memory_state` or move the `all_txns_count` check to transfer_leadership. I prefer to move the check.

bharathv commented:
After running some experiments (by @rystsov) it turns out that writing with `acks=all` for every replicate data request better amortizes the performance (without a prepared call), than writing with `acks=1` + prepared message with `acks=all` (has 2k ops better higher throughput). It has its own quirks but the side effect of implementing such a design is that we won't have in memory state on partitions that we need to checkpoint on failover. The code will be simpler and failover does not need separate handling. Looks like we are preferring that refactor, so we do not need this patch anymore, closing it out.
